### PR TITLE
Adding kubernetes deployment

### PR DESCRIPTION
**Pull request by blankdots**
Deployment of LocalEGA to Kubernetes/Openshift clusters includes:
* YAML files for a more hands-on deployment;
* a python deploy script using the https://github.com/kubernetes-client/python ;
* testing script in `deployments/kube/test`;
* Readme files on how to make use of the solution.

Missing: details added for http://localega.readthedocs.io/ - I think these should come after the PR is reviewed and approved.

> Do we need a travis stage to test this? 

Most likely yes. Example here: https://github.com/LiliC/travis-minikube/blob/master/.travis.yml but most likely we will need to adjust current tests.

> Where has this been tested?

So far only minikube, testing deployment on other Kubernetes/Openshift clusters is work in progress.

>  Will we have Horizontal Pod auto-scaling?

~Yes, if I am fast enough to test it and commit it.~ - Added simple autoscale based on [average CPU utilisation ](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/#create-horizontal-pod-autoscaler). If we need other metrics we need to change to `V2beta1` instead of `V1` API.

> How does one review this?

There are instructions in README or just ask for video demo.

### CODE REVIEW

File: deployments/kube/README.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+We provide an python script based on https://github.com/kubernetes-client/python that sets up all the necessary configuration (e.g. generating keys, certificates, configuration files etc.) and pods along with necessary services and volumes.
+The script is intended to work both with a minikube or any Kubernetes cluster, provided the user has an API key.
+
+**NOTES:**
+  - **Requires Python >3.6.**
+  - **Work in Progress**
+
+The script is in `auto` folder and can be run as:
+```
+cd ~/LocalEGA/deployment/kube/auto
```
Comment (dtitov) [inline deployments/kube/README.md]: It's `deployments`, plural.

File: deployments/kube/README.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+pip install -r requirements.txt
+kubectl create -f ../yml/cega-mq --namespace=testing
```
Comment (dtitov) [inline deployments/kube/README.md]: Doesn't work with `--namespace`...

File: deployments/kube/README.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+pyton deploy.py --config --deploy all
```
Comment (dtitov) [inline deployments/kube/README.md]: `python` 😸

File: deployments/kube/README.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+The script is intended to work both with a minikube or any Kubernetes cluster, provided the user has an API key.
+
+**NOTES:**
+  - **Requires Python >3.6.**
+  - **Work in Progress**
+
+The script is in `auto` folder and can be run as:
+```
+cd ~/LocalEGA/deployment/kube/auto
+pip install -r requirements.txt
```
Comment (dtitov) [inline deployments/kube/README.md]: On MacOS might be required to explicitly call `pip3`.

File: deployments/kube/README.md (hunk 3 above)
Comment (dtitov) [inline deployments/kube/README.md]: On MacOS might be required to explicitly call `python3`.

**Comment by dtitov** [comment]
I tested it with Minicube. Deployment works (haven't deployed `cega-users` though), but I didn't try the functionality itself (e.g. submission or ingestion). There are some small comments above and a major suggestion to include `cega-users` generation in the `deploy.py`

Another suggestion from me is to brainstorm which services do we want to have as **Deployments** and which as **Stateful sets**, because it appeared that it's not that obvious decision. For now, I guess, we can make them all to be Deployments (based on this article: https://akomljen.com/kubernetes-persistent-volumes-with-deployment-and-statefulset/) and then, after we discuss it, we can switch some of them to Stateful sets.

File: deployments/kube/README.md (hunk 2 above)
Comment (blankdots) [reply deployments/kube/README.md]: now it is `--ns` changed it ...forgot to update the docs

File: deployments/kube/README.md (hunk 4 above)
Comment (blankdots) [reply deployments/kube/README.md]: it is not for MacOS only ... it is how one has set up their python/python3 default/environments ... and it it specified at the beginning for the file that one needs Python > 3.6.3

File: deployments/kube/README.md (hunk 3 above)
Comment (blankdots) [reply deployments/kube/README.md]: will fix

Comment (blankdots) [reply deployments/kube/README.md]: same as for `pip3` - I cannot control what is on everyone's computer

**Comment by blankdots** [comment]
The purpose of the issue was to set up **LocalEGA** not to replicate the bootstrap from the docker-compose. Connection to CEGA should be known when setting up LocalEGA. I made `cega-mq` and `cega-users` just for testing purposes the same for the testing script, these could have well be missing from this PR.

This being said **I will add** fake-cega to deploy script (bear in mind the deploy script was just an idea, and it turned out to be a simple way).

Regarding StatefulSet vs Deployments I agree only with moving PostgresDB to a StatefulSet only if there is a clear requirement running it in a cluster like setup ... and of course do another PR with it.
For now it works and it is tested, if there are other requirements this means a new issue.

**Comment by dtitov** [comment]
I tested the new version (with `cega-users` in the `deploy.py`) and it works like a charm 👍

```
$ minikube service list
|-------------|----------------------|--------------------------------|
|  NAMESPACE  |         NAME         |              URL               |
|-------------|----------------------|--------------------------------|
| default     | kubernetes           | No node port                   |
| kube-system | kube-dns             | No node port                   |
| kube-system | kubernetes-dashboard | http://192.168.99.100:30000    |
| testing     | cega-mq              | http://192.168.99.100:32703    |
|             |                      | http://192.168.99.100:31314    |
|             |                      | http://192.168.99.100:31063    |
| testing     | cega-mq-management   | http://192.168.99.100:31668    |
| testing     | cega-users           | http://192.168.99.100:30250    |
| testing     | db                   | No node port                   |
| testing     | inbox                | http://192.168.99.100:31882    |
| testing     | keys                 | No node port                   |
| testing     | minio                | No node port                   |
| testing     | minio-service        | http://192.168.99.100:32650    |
| testing     | mq                   | No node port                   |
| testing     | mq-management        | http://192.168.99.100:30669    |
|-------------|----------------------|--------------------------------|
```

**Comment by dtitov** [comment]
Sucessfully tested it manually ✅

P.S. Just add the `--stableID` before `EGAF$(shell cat $(FILE).md5)` in the test Makefile.

**Comment by dtitov** [comment]
LGTM.

**Pull request merged**